### ISSUE TITLE
Tips for Termux installation

### ISSUE THREAD

**Issue by Manamama** [google/sentencepiece#950]
This had to be fixed in Termux: 

/data/data/com.termux/files/home/downloads/sentencepiece/CMakeLists.txt

1. `set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -llog")`: This line is setting a CMake variable called `CMAKE_SHARED_LINKER_FLAGS`, which specifies the flags to be used by the linker when creating shared libraries⁵. The `-llog` flag tells the linker to link against `liblog.so`, which provides the `__android_log_write` function that was causing the undefined reference error in your initial build attempt⁶.

2. `cmake .. -DCMAKE_INSTALL_PREFIX=$PREFIX`: Here, you're invoking CMake and passing it a command-line argument to set the `CMAKE_INSTALL_PREFIX` variable. This variable determines the directory where the project will be installed[^10^]¹¹. By setting it to `$PREFIX`, you're telling CMake to install the project in the directory specified by the `PREFIX` environment variable, which in Termux is typically `/data/data/com.termux/files/usr`.

After applying these fixes and successfully building the project: 
```
cd ..
cd python/
python setup.py bdist_wheel
pip install dist/sentencepiece*.whl
pip show sentencepiece
```

**Comment by FindingMeaning** [comment]
😍😭❤️ you're a genius I was about to give up ,  please please please can you explain this in more detail I am using tarfile to install this and I cannot find the python/ directory 😢 and I am a little bit confused about where I've to make that edit is that line number specific in that file ?

**Comment by Manamama** [comment]
I am not a genius, by far. I run most by Ms Bing (mostly, Precise mode) or Claude AI, with my neofetch results and just ask it politely, yet insistently, and interactively. 

Re your question - it requires 
```
github clone https://github.com/google/sentencepiece
cd sentencepiece
```

first, obviously, then changing these tidbits ( _in .../sentencepiece/CMakeLists.txt_, as I wrote) that you will see, and maybe paying attention to the missing libraries that 
`python setup ... `
or
`pip install ... `
logs may throw at you, and installing these missing bits (modules etc.), if any. 

In short - just paste these errors to a live session with any LLM/AI and explain to it politely what you are after, assuming it is also a blind man seeing (your) elephant, that is the system errors.

**Comment by Manamama** [comment]
Another tip for Termux lovers (like me)

```
~ $ pkg list-all | grep gperf

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

gperf/stable,now 3.1-7 aarch64 [installed,automatic]
~ $ 
```

or 

```
WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

cmake-curses-gui/stable 3.28.1 aarch64
cmake/stable,now 3.28.1 aarch64 [installed]
extra-cmake-modules/x11 5.112.0 aarch64
~ $ 
```
is what is needed to compile, the termux equivalents thereof 

Anyhow, use that syntax to find the Termux precompiled equivalents of most anything: numpy (needed for numba), scikit, torch, etc. 


Another trick that (may, a big may) work out of the box

```
~ $ time  pip install --no-dependencies   --no-binary SentencePiece SentencePiece
Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com
Requirement already satisfied: SentencePiece in /data/data/com.termux/files/usr/lib/python3.11/site-packages (0.2.0)

real	0m2.616s
user	0m1.569s
sys	0m0.577s
~ $ 

```

**Comment by Manamama** [comment]
FYI, I know it should not be here, as it is not a bug, but I wrote [a quick script](https://github.com/Manamama/AI_scripts_Public_1/blob/main/SentencePiece_a_file.py) that runs it in Termux etc, with [cute statistics](https://gist.github.com/Manamama/bb4d360b9b337cf926cd68f00e22f444). 

Ms Bing explained it to me via PlantUML as below
![SentencePiece](https://gist.github.com/assets/78492008/1575d0f7-9b38-42e4-b533-6397f8f6c9b0)

**Comment by FindingMeaning** [comment]
> first, obviously, then changing these tidbits ( _in .../sentencepiece/CMakeLists.txt_, as I wrote) that you will see, and maybe paying attention to the missing libraries that 
> `python setup ... `
> or
> `pip install ... `
> logs may throw at you, and installing these missing bits (modules etc.), if any. 

Thankyou very very much 😌 I've successfully installed sentence-transformers because of you, for me you'll always be a genius and a person who helped me when I was loosing hope 😭

![RDT_20231219_084911264719658203460677.jpg](https://github.com/google/sentencepiece/assets/146155413/bea5480e-b47b-4bec-bb57-58069e269be2)

Thank you very very very much once again ☺️❤️

**Comment by xhy2008** [comment]
Which version of sentencepiece are you using?
I looked through the CMakeLists.txt,but I didn't find the line you mentioned.
I am using the latest version,and it still make errors.

**Comment by FindingMeaning** [comment]
Just add this line at top of CMakeLists.txt `set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -llog")`
if you still have problems see this [Reddit Post](https://www.reddit.com/r/termux/comments/18ojp5j/how_to_install_sentencetransformers_with_all/?utm_source=share&utm_medium=mweb3x&utm_name=mweb3xcss&utm_term=1&utm_content=share_button).

**Comment by taku910** [comment]
Let me close this issue, as it seems to be already resolved.

**Comment by cargilcm** [comment]
CFLAGS="-llog" CXXFLAGS="-llog" LDFLAGS="-llog" pip install sentencepiece
